bitsail-connectors/connector-redis/src/main/java/com/bytedance/bitsail/connector/redis/sink/RedisWriter.java

```diff
@@ -0,0 +1,249 @@
+      throw new BitSailException(CommonErrorCode.CONVERT_NOT_SUPPORT,
+          String.format("The score can't convert to double. And the score is %s.",
+              new String(scoreInBytes)));
+    }
+  }
+
+  @Override
+  @SneakyThrows
+  public void flush(boolean endOfInput) throws IOException {
+    processorId++;
+    try (PipelineProcessor processor = genPipelineProcessor(recordQueue.size(), this.complexTypeWithTtl)) {
+      Row record;
+      while ((record = recordQueue.poll()) != null) {
+
+        //***************** todo: remove this logic *****************
+        byte[] key = ((String) record.getField(0)).getBytes();
+        byte[] value = ((String) record.getField(1)).getBytes();
+        byte[] scoreOrHashKey = value;
+        if (columnSize == SORTED_SET_OR_HASH_COLUMN_SIZE) {
+          value = ((String) record.getField(2)).getBytes();
+          // Replace empty key with additionalKey in sorted set and hash.
+          if (key.length == 0) {
+            key = commandDescription.getAdditionalKey().getBytes();
+          }
+        }
+        //*******************************************************
+
+        if (commandDescription.getJedisCommand() == JedisCommand.ZADD) {
+          // sortedSet
+          processor.addInitialCommand(new Command(commandDescription, key, parseScoreFromBytes(scoreOrHashKey), value));
+        } else if (commandDescription.getJedisCommand() == JedisCommand.HSET) {
+          // hash
+          processor.addInitialCommand(new Command(commandDescription, key, scoreOrHashKey, value));
+        } else {
+          // set and string
+          processor.addInitialCommand(new Command(commandDescription, key, value));
+        }
+      }
+      retryer.call(processor::run);
+
+    } catch (ExecutionException | RetryException e) {
+      if (e.getCause() instanceof BitSailException) {
+        throw (BitSailException) e.getCause();
+      } else if (e.getCause() instanceof UnexpectedException) {
+        throw (UnexpectedException) e.getCause();
+      }
+      throw e;
+    } catch (IOException e) {
+      throw new RuntimeException("Error while init jedis client.", e);
+    }
+  }
+
+  /**
+   * Gen pipeline processor for jedis commands.
+   *
+   * @return Jedis pipeline.
+   */
+  protected PipelineProcessor genPipelineProcessor(int commandSize, boolean complexTypeWithTtl) throws ExecutionException, RetryException {
+    return new RedisPipelineProcessor(jedisPool, jedisFetcher, commandSize, processorId, logSampleInterval, complexTypeWithTtl);
+  }
+
+  @Override
+  public List<CommitT> prepareCommit() throws IOException {
+    return null;
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      if (CollectionUtils.isNotEmpty(recordQueue)) {
+        flush(true);
+      }
+    } catch (IOException e) {
+      LOG.error("flush the last pipeline occurs error.", e);
+      throw e;
+    } finally {
+      Writer.super.close();
```

Comment:
what does this do

Reply:
the record queue is a cache to cache record to the batchInterval and flush it. when at the last, before close, we will check if there is some record in record queue that don't reach batchInterval. we need to flush it.